metrics/firebase/functions/test/main_test.dart

```diff
@@ -0,0 +1,250 @@
+    final _documentReferenceMock = DocumentReferenceMock();
+    final _documentSnapshotMock = DocumentSnapshotMock();
+    final _eventContextMock = EventContextMock();
+
+    tearDown(() {
+      reset(_documentSnapshotMock);
+      reset(_eventContextMock);
+      reset(_firestoreMock);
+      reset(_collectionReferenceMock);
+      reset(_documentReferenceMock);
+    });
+
+    PostExpectation<Firestore> whenFirestore() {
+      return when(_documentSnapshotMock.firestore);
+    }
+
+    PostExpectation<CollectionReference> whenCollection(String withName) {
+      whenFirestore().thenReturn(_firestoreMock);
+
+      return when(_firestoreMock.collection(withName));
+    }
+
+    PostExpectation<DocumentReference> whenDocument({
+      String withCollectionName,
+    }) {
+      whenCollection(withCollectionName).thenReturn(_collectionReferenceMock);
+
+      return when(_collectionReferenceMock.document(any));
+    }
+
+    PostExpectation<Future<void>> whenDocumentData({
+      String withCollectionName,
+    }) {
+      whenDocument(
+        withCollectionName: withCollectionName,
+      ).thenReturn(_documentReferenceMock);
+
+      return when(_documentReferenceMock.setData(any, any));
+    }
+
+    PostExpectation whenSnapshotData() {
+      return when(_documentSnapshotMock.data);
+    }
+
+    test(
+      "sets a build day document with the zero duration if the build document snapshot's duration is a null",
+      () async {
+        whenDocumentData(withCollectionName: buildDaysCollectionName)
+            .thenAnswer((_) => Future.value());
+
+        final Map<String, dynamic> buildWithoutDuration = Map.from(build);
+        buildWithoutDuration.remove('duration');
+
+        whenSnapshotData()
+            .thenReturn(DocumentData.fromMap(buildWithoutDuration));
+
+        await onBuildAddedHandler(_documentSnapshotMock, _eventContextMock);
+
+        final documentDataMatcher = predicate<DocumentData>((data) {
+          return data.getNestedData('totalDuration').getInt('operand') == 0;
+        });
+
+        final optionMatcher = predicate<SetOptions>((option) => option.merge);
+
+        verify(
+          _documentReferenceMock.setData(
+            argThat(documentDataMatcher),
+            argThat(optionMatcher),
+          ),
+        ).called(1);
+      },
+    );
+
+    test(
+      "uses a composite document id for the build days collection",
+      () async {
+        whenDocument(withCollectionName: buildDaysCollectionName)
+            .thenReturn(_documentReferenceMock);
+        whenSnapshotData().thenReturn(DocumentData.fromMap(build));
+
+        await onBuildAddedHandler(_documentSnapshotMock, _eventContextMock);
+
+        final documentId = '${projectId}_${utcDate.millisecondsSinceEpoch}';
+
+        verify(_collectionReferenceMock.document(documentId)).called(1);
+      },
+    );
+
+    test(
+      "creates a build days document from the build document snapshot's data",
```

Comment:
I believe it would be better to test each `BuildDayData` field in a separate test. What do you think?